src/col/vct/col/ast/lang/pvl/PVLInvocationImpl.scala

```diff
@@ -1,17 +1,30 @@
 package vct.col.ast.lang.pvl
 
-import vct.col.ast.{PVLInvocation, TProcess, TResource, Type}
-import vct.col.print.{Ctx, Doc, DocUtil, Empty, Text, Group}
+import vct.col.ast.{Applicable, PVLInvocation, TClass, TProcess, TResource, Type}
+import vct.col.print.{Ctx, Doc, DocUtil, Empty, Group, Text}
 import vct.col.resolve.ctx._
 import vct.col.ast.ops.PVLInvocationOps
+import vct.col.ref.Ref
 
 trait PVLInvocationImpl[G] extends PVLInvocationOps[G] { this: PVLInvocation[G] =>
   override lazy val t: Type[G] = ref.get match {
     case RefFunction(decl) => decl.returnType.particularize(decl.typeArgs.zip(typeArgs).toMap)
     case RefProcedure(decl) => decl.returnType
     case RefPredicate(_) => TResource()
-    case RefInstanceFunction(decl) => decl.returnType.particularize(decl.typeArgs.zip(typeArgs).toMap)
-    case RefInstanceMethod(decl) => decl.returnType
+    case RefInstanceFunction(decl) =>
+      val returnType = decl.returnType.particularize(decl.typeArgs.zip(typeArgs).toMap)
+      obj.flatMap { e =>
+        e.t.asClass.map { tcls =>
+          tcls.instantiate(returnType)
+        }
+      }.getOrElse(returnType)
+    case RefInstanceMethod(decl) =>
+      val returnType = decl.returnType.particularize(decl.typeArgs.zip(typeArgs).toMap)
+      obj.flatMap { e =>
+        e.t.asClass.map { tcls =>
+          tcls.instantiate(returnType)
+        }
+      }.getOrElse(returnType)
```

Comment:
You don't need to be conservative here: `t` may not be called before the node is checked, so please use `e.t.asClass.get`.

Reply:
Also if the `flatMap` is for the case that the object is implicitly `this`, maybe put that in a comment :)